File: internal/types/configuration_drift.go
package types

import (
	"encoding/json"

	"github.com/google/uuid"
)

type ConfigurationDrift struct {
	DeploymentID           uuid.UUID                           `json:"deploymentId"`
	ApplicationID          uuid.UUID                           `json:"applicationId"`
	HasDrift               bool                                `json:"hasDrift"`
	NewRequiredVariables   []ConfigurationDriftVariable        `json:"newRequiredVariables"`
	MissingVariables       []ConfigurationDriftVariable        `json:"missingVariables"`
	RemovedVariables       []ConfigurationDriftRemovedValue    `json:"removedVariables"`
	TypeChanges            []ConfigurationDriftTypeChange      `json:"typeChanges"`
	DefaultChanges         []ConfigurationDriftDefaultChange   `json:"defaultChanges"`
	SecretReferenceChanges []ConfigurationDriftReferenceChange `json:"secretReferenceChanges"`
}

type ConfigurationDriftVariable struct {
	Key           string                   `json:"key"`
	Type          VariableType             `json:"type"`
	IsRequired    bool                     `json:"isRequired"`
	Source        VariableResolutionSource `json:"source"`
	Value         json.RawMessage          `json:"value,omitempty"`
	ReferenceID   string                   `json:"referenceId,omitempty"`
	ReferenceName string                   `json:"referenceName,omitempty"`
	Redacted      bool                     `json:"redacted"`
}

type ConfigurationDriftRemovedValue struct {
	Key string `json:"key"`
}

type ConfigurationDriftTypeChange struct {
	Key          string       `json:"key"`
	ExpectedType VariableType `json:"expectedType"`
	DeployedType string       `json:"deployedType"`
}

type ConfigurationDriftDefaultChange struct {
	Key           string          `json:"key"`
	Type          VariableType    `json:"type"`
	CurrentValue  json.RawMessage `json:"currentValue,omitempty"`
	DeployedValue json.RawMessage `json:"deployedValue,omitempty"`
}

type ConfigurationDriftReferenceChange struct {
	Key           string       `json:"key"`
	Type          VariableType `json:"type"`
	ReferenceID   string       `json:"referenceId,omitempty"`
	ReferenceName string       `json:"referenceName,omitempty"`
	Redacted      bool         `json:"redacted"`
}

File: frontend/ui/src/app/types/configuration-drift.ts
import {VariableType} from './variable-set';

export interface ConfigurationDrift {
  deploymentId: string;
  applicationId: string;
  hasDrift: boolean;
  newRequiredVariables: ConfigurationDriftVariable[];
  missingVariables: ConfigurationDriftVariable[];
  removedVariables: ConfigurationDriftRemovedValue[];
  typeChanges: ConfigurationDriftTypeChange[];
  defaultChanges: ConfigurationDriftDefaultChange[];
  secretReferenceChanges: ConfigurationDriftReferenceChange[];
}

export interface ConfigurationDriftVariable {
  key: string;
  type: VariableType;
  isRequired: boolean;
  source: string;
  value?: unknown;
  referenceId?: string;
  referenceName?: string;
  redacted?: boolean;
}

export interface ConfigurationDriftRemovedValue {
  key: string;
}

export interface ConfigurationDriftTypeChange {
  key: string;
  expectedType: VariableType;
  deployedType: string;
}

export interface ConfigurationDriftDefaultChange {
  key: string;
  type: VariableType;
  currentValue?: unknown;
  deployedValue?: unknown;
}

export interface ConfigurationDriftReferenceChange {
  key: string;
  type: VariableType;
  referenceId?: string;
  referenceName?: string;
  redacted: boolean;
}
